dreammall-earth/dreammall.earth
# feat(other): check-in encrypted tfstate

Motivation
----------
Terraform/Tofu needs state to manage infrastructure and configuration:
https://developer.hashicorp.com/terraform/language/state

I found out that I cannt scale the cluster anymore, if I loose the
`terraform.tfstate` file. Unfortunately, it contains our `kubeconfig`
in clear.

So I followed [OpenTofu's docs](https://opentofu.org/docs/language/state/encryption/)
how to encrypt the tfstate. As symmetric key, I re-used the
`hcloud_token` which is required already.

How to test
-----------
1. See encrypted data in `./infrastructure/kube-hetzner/terraform.tfstate`
2. If you do `tofu output --raw kubeconfig` it will output:
```
╷
│ Error: Unable to build encryption key data
│
│ key_provider.pbkdf2.hcloud_token failed with error: passphrase is too short (minimum 16 characters)
```
2. But if you set `export TF_VAR_hcloud_token=<hcloud_token>`:
3. `tofu output --raw kubeconfig` and it will output:
```
apiVersion: v1
clusters:
- cluster:
    certific
[...]
```
